> Would be good to run the end2end performance regression pipeline and see the impact on performance. There's more threads now, let's see how it does.

Sounds good.

@DrJessop what do you run for e2e perf for resnet and bert? I tried this locally but even on main it failed the check:
```
pytest "models/demos/resnet/tests/test_perf_resnet.py::test_perf_bare_metal[8-0.015-25]"
E       AssertionError: resnet50 224x224_batchsize8 inference is too slow
E       assert 0.21947139104207355 < 0.015
```

--- a/tt_metal/impl/dispatch/command_queue.cpp
+++ b/tt_metal/impl/dispatch/command_queue.cpp
@@ -1110,7 +1110,7 @@ void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Bu
 
 void EnqueueReadBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking) {
     detail::DispatchStateCheck(true);
-    cq.run_command(CommandQueueInterface{
+    cq.run_command(CommandInterface{
         .type = EnqueueCommandType::ENQUEUE_READ_BUFFER,
         .blocking = blocking,
         .buffer = buffer,
@@ -1123,7 +1123,7 @@ void EnqueueReadBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper
         using T = std::decay_t<decltype(b)>;
         std::shared_future<void> f;
         if constexpr (std::is_same_v<T, std::reference_wrapper<Buffer>> || std::is_same_v<T, std::shared_ptr<Buffer> > ) {
-            cq.device()->hw_command_queue(cq.id()).enqueue_read_buffer(b, dst, blocking);
+            cq.hw_command_queue().enqueue_read_buffer(b, dst, blocking);
         }
     }, buffer);
 }
@@ -1131,7 +1131,7 @@ void EnqueueReadBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper
 void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer,
                                           const void* src, bool blocking) {
     detail::DispatchStateCheck(true);
-    cq.run_command(CommandQueueInterface{
+    cq.run_command(CommandInterface{
         .type = EnqueueCommandType::ENQUEUE_WRITE_BUFFER,
         .blocking = blocking,
         .buffer = buffer,
@@ -1143,10 +1143,8 @@ void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrappe
                                           const void* src, bool blocking) {
     std::visit ( [&cq, src, blocking](auto&& b) {
         using T = std::decay_t<decltype(b)>;
-        Device * device = cq.device();
-        auto cq_id = cq.id();
         if constexpr (std::is_same_v<T, std::reference_wrapper<Buffer>> || std::is_same_v<T, std::shared_ptr<Buffer>> ) {
-            device->hw_command_queue(cq_id).enqueue_write_buffer(b, src, blocking);
+            cq.hw_command_queue().enqueue_write_buffer(b, src, blocking);
         }
     }, buffer);
 }
@@ -1154,7 +1152,7 @@ void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrappe
 void EnqueueProgram(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace) {
     detail::DispatchStateCheck(true);
     TT_ASSERT(cq.id() == 0, "EnqueueProgram only supported on first command queue on device for time being.");
-    cq.run_command(CommandQueueInterface{
+    cq.run_command(CommandInterface{
         .type = EnqueueCommandType::ENQUEUE_PROGRAM,
         .blocking = blocking,
         .program = program,
@@ -1168,39 +1166,37 @@ void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<
         ZoneScoped;
         using T = std::decay_t<decltype(program)>;
         Device * device = cq.device();
-        auto cq_id = cq.id();
         if constexpr (std::is_same_v<T, std::reference_wrapper<Program>>) {
             detail::CompileProgram(device, program);
             program.get().allocate_circular_buffers();
             detail::ValidateCircularBufferRegion(program, device);
-            device->hw_command_queue(cq_id).enqueue_program(program, trace, blocking);
+            cq.hw_command_queue().enqueue_program(program, trace, blocking);
         } else if constexpr (std::is_same_v<T, std::shared_ptr<Program>>) {
             detail::CompileProgram(device, *program);
             program->allocate_circular_buffers();
             detail::ValidateCircularBufferRegion(*program, device);
-            device->hw_command_queue(cq_id).enqueue_program(*program, trace, blocking);
+            cq.hw_command_queue().enqueue_program(*program, trace, blocking);
         }
     }, program);
 }
 
 void Finish(CommandQueue& cq) {
     detail::DispatchStateCheck(true);
-    cq.run_command(CommandQueueInterface{
+    cq.run_command(CommandInterface{
         .type = EnqueueCommandType::FINISH,
         .blocking = true
     });
 }
 
 void FinishImpl(CommandQueue& cq) {
-    cq.device()->hw_command_queue(cq.id()).finish();
+    cq.hw_command_queue().finish();
 }
 
 
 Trace BeginTrace(CommandQueue& cq) {
     // Resets the command queue state
-    Device * device = cq.device();
-    device->hw_command_queue(cq.id()).restart();
-    return Trace(device->hw_command_queue(cq.id()));
+    cq.hw_command_queue().restart();
+    return Trace(cq.hw_command_queue());
 }
 
 void EndTrace(Trace& trace) {
@@ -1237,14 +1233,12 @@ namespace detail {
 void EnqueueRestart(CommandQueue& cq) {
     ZoneScoped;
     detail::DispatchStateCheck(true);
-    cq.device()->hw_command_queue(cq.id()).restart();
+    cq.hw_command_queue().restart();
 }
 
 }
 
 CommandQueue::CommandQueue(Device* device, uint32_t id) : device_ptr(device), cq_id(id) {
-    bool value = tt::parse_env("TT_METAL_ASYNC_QUEUES", false);
-    mode = value ? CommandQueueMode::ASYNC : CommandQueueMode::PASSTHROUGH;
     if (async_mode()) {
         start_worker();
     }
@@ -1254,22 +1248,27 @@ CommandQueue::~CommandQueue() {
     if (async_mode()) {
         stop_worker();
     }
-    TT_ASSERT(worker_queue.empty(), "CQ{} worker queue must be empty on destruction", cq_id);
+    TT_ASSERT(this->worker_queue.empty(), "CQ{} worker queue must be empty on destruction", cq_id);
+}
+
+HWCommandQueue& CommandQueue::hw_command_queue() {
+    return device()->hw_command_queue(this->cq_id);
 }
 
 void CommandQueue::wait_until_empty() {
-    log_trace(tt::LogDispatch, "CQ{} WFI start", cq_id);
-    // Insert a token command to flush all prior commands
-    worker_queue.push(CommandQueueInterface{
-        .type = EnqueueCommandType::INVALID
+    log_trace(LogDispatch, "CQ{} WFI start", cq_id);
+    // Insert a flush token to push all prior commands to completion
+    // Necessary to avoid implementing a peek and pop on the lock-free queue
+    this->worker_queue.push(CommandInterface{
+        .type = EnqueueCommandType::FLUSH
     });
     while (true) {
-        if (worker_queue.empty()) {
+        if (this->worker_queue.empty()) {
             break;
         }
         std::this_thread::sleep_for(std::chrono::microseconds(10));
     }
-    log_trace(tt::LogDispatch, "CQ{} WFI complete", cq_id);
+    log_trace(LogDispatch, "CQ{} WFI complete", cq_id);
 }
 
 void CommandQueue::set_mode(const CommandQueueMode& mode_) {
@@ -1283,44 +1282,44 @@ void CommandQueue::set_mode(const CommandQueueMode& mode_) {
 }
 
 void CommandQueue::start_worker() {
-    if (worker_state == CommandQueueState::RUNNING) {
+    if (this->worker_state == CommandQueueState::RUNNING) {
         return;  // worker already running, exit
     }
-    worker_state = CommandQueueState::RUNNING;
-    worker_thread = std::make_unique<std::thread>(std::thread(&CommandQueue::run_worker, this));
+    this->worker_state = CommandQueueState::RUNNING;
+    this->worker_thread = std::make_unique<std::thread>(std::thread(&CommandQueue::run_worker, this));
     tt::log_debug(tt::LogDispatch, "CQ{} started worker thread", cq_id);
 }
 
 void CommandQueue::stop_worker() {
-    if (worker_state == CommandQueueState::IDLE) {
+    if (this->worker_state == CommandQueueState::IDLE) {
         return;  // worker already stopped, exit
     }
-    worker_state = CommandQueueState::TERMINATE;
-    worker_thread->join();
-    worker_state = CommandQueueState::IDLE;
+    this->worker_state = CommandQueueState::TERMINATE;
+    this->worker_thread->join();
+    this->worker_state = CommandQueueState::IDLE;
     tt::log_debug(tt::LogDispatch, "CQ{} stopped worker thread", cq_id);
 }
 
 void CommandQueue::run_worker() {
     // forever loop checking for commands in the worker queue
     while (true) {
-        if (worker_queue.empty()) {
-            if (worker_state == CommandQueueState::TERMINATE) {
+        if (this->worker_queue.empty()) {
+            if (this->worker_state == CommandQueueState::TERMINATE) {
                 break;
             }
             std::this_thread::yield();
         } else {
-            auto command = worker_queue.pop();
+            auto command = this->worker_queue.pop();
             run_command_impl(*command);
         }
     }
 }
 
-void CommandQueue::run_command(const CommandQueueInterface& command) {
-    log_trace(tt::LogDispatch, "CQ{} received {} in {} mode", cq_id, command.type, async_mode() ? "ASYNC" : "PASSTHROUGH");
+void CommandQueue::run_command(const CommandInterface& command) {
+    log_trace(LogDispatch, "CQ{} received {} in {} mode", cq_id, command.type, async_mode() ? "ASYNC" : "PASSTHROUGH");
     if (async_mode()) {
-        worker_queue.push(command);
-        if (command.blocking.has_value() && *command.blocking == true) {
+        this->worker_queue.push(command);
+        if (command.blocking.has_value() and *command.blocking == true) {
             wait_until_empty();
         }
     } else {
@@ -1328,8 +1327,8 @@ void CommandQueue::run_command(const CommandQueueInterface& command) {
     }
 }
 
-void CommandQueue::run_command_impl(const CommandQueueInterface& command) {
-    log_trace(tt::LogDispatch, "CQ{} running {}", cq_id, command.type);
+void CommandQueue::run_command_impl(const CommandInterface& command) {
+    log_trace(LogDispatch, "CQ{} running {}", cq_id, command.type);
     switch (command.type) {
         case EnqueueCommandType::ENQUEUE_READ_BUFFER:
             TT_ASSERT(command.dst.has_value(), "Must provide a dst!");
@@ -1351,12 +1350,13 @@ void CommandQueue::run_command_impl(const CommandQueueInterface& command) {
         case EnqueueCommandType::FINISH:
             FinishImpl(*this);
             break;
-        case EnqueueCommandType::INVALID:
+        case EnqueueCommandType::FLUSH:
+            // Used by CQ to push prior commands
             break;
         default:
             TT_THROW("Invalid command type");
     }
-    // log_trace(tt::LogDispatch, "CQ{} running {} complete", cq_id, command.type);
+    // log_trace(LogDispatch, "CQ{} running {} complete", cq_id, command.type);
 }
 
 }  // namespace tt::tt_metal

--- a/tt_metal/impl/dispatch/command_queue.cpp
+++ b/tt_metal/impl/dispatch/command_queue.cpp
@@ -6,10 +6,12 @@
 
 #include <algorithm>  // for copy() and assign()
 #include <iterator>   // for back_inserter
+#include <memory>
 
 #include "debug_tools.hpp"
 #include "dev_msgs.h"
 #include "llrt/watcher.hpp"
+#include "logger.hpp"
 #include "noc/noc_parameters.h"
 #include "tt_metal/detail/program.hpp"
 #include "tt_metal/detail/tt_metal.hpp"
@@ -1108,6 +1110,15 @@ void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Bu
 
 void EnqueueReadBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking) {
     detail::DispatchStateCheck(true);
+    cq.run_command(CommandQueueInterface{
+        .type = EnqueueCommandType::ENQUEUE_READ_BUFFER,
+        .blocking = blocking,
+        .buffer = buffer,
+        .dst = dst
+    });
+}
+
+void EnqueueReadBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking) {
     std::visit ( [&cq, dst, blocking](auto&& b) {
         using T = std::decay_t<decltype(b)>;
         std::shared_future<void> f;
@@ -1124,6 +1135,16 @@ void EnqueueReadBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buf
 void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer,
                                           const void* src, bool blocking) {
     detail::DispatchStateCheck(true);
+    cq.run_command(CommandQueueInterface{
+        .type = EnqueueCommandType::ENQUEUE_WRITE_BUFFER,
+        .blocking = blocking,
+        .buffer = buffer,
+        .src = src
+    });
+}
+
+void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer,
+                                          const void* src, bool blocking) {
     std::visit ( [&cq, src, blocking](auto&& b) {
         using T = std::decay_t<decltype(b)>;
         // std::shared_future<void> f;
@@ -1141,9 +1162,18 @@ void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Bu
 }
 
 void EnqueueProgram(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace) {
-    ZoneScoped;
-    TT_ASSERT(cq.id() == 0, "EnqueueProgram only supported on first command queue on device for time being.");
     detail::DispatchStateCheck(true);
+    TT_ASSERT(cq.id() == 0, "EnqueueProgram only supported on first command queue on device for time being.");
+    cq.run_command(CommandQueueInterface{
+        .type = EnqueueCommandType::ENQUEUE_PROGRAM,
+        .blocking = blocking,
+        .program = program,
+        .trace = trace
+    });
+}
+
+void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace) {
+    ZoneScoped;
     std::visit ( [&cq, blocking, trace](auto&& program) {
         ZoneScoped;
         using T = std::decay_t<decltype(program)>;
@@ -1173,6 +1203,12 @@ void EnqueueProgram(CommandQueue& cq, std::variant < std::reference_wrapper<Prog
 
 void Finish(CommandQueue& cq) {
     detail::DispatchStateCheck(true);
+    cq.run_command(CommandQueueInterface{
+        .type = EnqueueCommandType::FINISH
+    });
+}
+
+void FinishImpl(CommandQueue& cq) {
     // std::shared_future<void> f;
     // cq.submit( [device = cq.device(), cq_id = cq.id()] {
         cq.device()->hw_command_queue(cq.id()).finish();
@@ -1233,4 +1269,127 @@ void EnqueueRestart(CommandQueue& cq) {
 
 }
 
+CommandQueue::CommandQueue(Device* device, uint32_t id) : device_ptr(device), cq_id(id) {
+    if (async_mode) {
+        start_worker();
+    }
+    tt::log_debug(tt::LogDispatch, "CQ{} async mode = {}", cq_id, async_mode ? "ENABLED" : "DISABLED");
+    TT_ASSERT(worker_queue.empty(), "CQ{} worker queue must be empty on construction", cq_id);
+}
+
+CommandQueue::~CommandQueue() {
+    if (async_mode) {
+        stop_worker();
+    }
+    TT_ASSERT(worker_queue.empty(), "CQ{} worker queue must be empty on destruction", cq_id);
+}
+
+void CommandQueue::wait_until_empty() {
+    if (async_mode) {
+        worker_queue.push(CommandQueueInterface{
+            .type = EnqueueCommandType::INVALID,
+            .token = true
+        });
+    }
+    while (true) {
+        if (worker_queue.empty()) {
+            break;
+        }
+        std::this_thread::sleep_for(std::chrono::microseconds(10));
+    }
+}
+
+void CommandQueue::start_worker() {
+    if (worker_state == CommandQueueState::RUNNING) {
+        return;  // worker already running, exit
+    }
+    worker_state = CommandQueueState::RUNNING;
+    worker_thread = std::make_unique<std::thread>(std::thread(&CommandQueue::run_worker, this));
+    tt::log_debug(tt::LogDispatch, "CQ{} started worker thread", cq_id);
+}
+
+void CommandQueue::stop_worker() {
+    if (worker_state == CommandQueueState::IDLE) {
+        return;  // worker already stopped, exit
+    }
+    worker_state = CommandQueueState::TERMINATE;
+    worker_thread->join();
+    worker_state = CommandQueueState::IDLE;
+    tt::log_debug(tt::LogDispatch, "CQ{} stopped worker thread", cq_id);
+}
+
+void CommandQueue::run_worker() {
+    // forever loop checking for commands in the worker queue
+    while (true) {
+        // if (worker_queue.empty()) {
+        //     std::this_thread::sleep_for(std::chrono::microseconds(10));
+        //     continue;
+        // } else {
+        //     run_command_impl(*worker_queue.pop());
+        // }
+        if (worker_queue.empty()) {
+            // log_info("CQ{} worker queue is empty", cq_id);
+            if (worker_state == CommandQueueState::TERMINATE) {
+                break;
+            }
+        } else {
+            auto command = worker_queue.pop();
+            if (command->token.has_value()) {
+                // acts like a flush
+            } else {
+                run_command_impl(*command);
+            }
+        }
+    }
+}
+
+void CommandQueue::run_command(const CommandQueueInterface& command) {
+    tt::log_debug(tt::LogDispatch, "CQ{} received command {} in {} mode", cq_id, command.type, async_mode ? "ASYNC" : "NORMAL");
+    if (async_mode) {
+        worker_queue.push(command);
+    } else {
+        run_command_impl(command);
+    }
+}
+
+void CommandQueue::run_command_impl(const CommandQueueInterface& command) {
+    tt::log_debug(tt::LogDispatch, "CQ{} running command {}", cq_id, command.type);
+    switch (command.type) {
+        case EnqueueCommandType::ENQUEUE_READ_BUFFER:
+            TT_ASSERT(command.dst.has_value(), "Must provide a dst!");
+            TT_ASSERT(command.buffer.has_value(), "Must provide a buffer!");
+            TT_ASSERT(command.blocking.has_value(), "Must specify blocking value!");
+            EnqueueReadBufferImpl(*this, command.buffer.value(), command.dst.value(), command.blocking.value());
+            break;
+        case EnqueueCommandType::ENQUEUE_WRITE_BUFFER:
+            TT_ASSERT(command.src.has_value(), "Must provide a src!");
+            TT_ASSERT(command.buffer.has_value(), "Must provide a buffer!");
+            TT_ASSERT(command.blocking.has_value(), "Must specify blocking value!");
+            EnqueueWriteBufferImpl(*this, command.buffer.value(), command.src.value(), command.blocking.value());
+            break;
+        case EnqueueCommandType::ENQUEUE_PROGRAM:
+            TT_ASSERT(command.program.has_value(), "Must provide a program!");
+            TT_ASSERT(command.blocking.has_value(), "Must specify blocking value!");
+            EnqueueProgramImpl(*this, command.program.value(), command.blocking.value(), command.trace);
+            break;
+        case EnqueueCommandType::FINISH:
+            tt::log_debug(tt::LogDispatch, "CQ{} received FINISH", cq_id);
+            FinishImpl(*this);
+            break;
+        default:
+            TT_THROW("Invalid command type");
+    }
+}
+
 }  // namespace tt::tt_metal
+
+std::ostream& operator<<(std::ostream& os, EnqueueCommandType const& type) {
+    switch (type) {
+        case EnqueueCommandType::ENQUEUE_READ_BUFFER: os << "ENQUEUE_READ_BUFFER"; break;
+        case EnqueueCommandType::ENQUEUE_WRITE_BUFFER: os << "ENQUEUE_WRITE_BUFFER"; break;
+        case EnqueueCommandType::ENQUEUE_PROGRAM: os << "ENQUEUE_PROGRAM"; break;
+        case EnqueueCommandType::FINISH: os << "FINISH"; break;
+        default: tt::log_fatal("Invalid command type!");
+    }
+    return os;
+}

--- a/tt_metal/impl/dispatch/command_queue.hpp
+++ b/tt_metal/impl/dispatch/command_queue.hpp
@@ -534,7 +534,8 @@ class CommandQueue {
     bool passthrough_mode() { return this->mode == CommandQueueMode::PASSTHROUGH; }
 
     static CommandQueueMode get_mode() {
-        int value = parse_env<int>("TT_METAL_ASYNC_QUEUES", static_cast<int>(CommandQueueMode::PASSTHROUGH));
+        // Envvar is used for bringup and debug only. Will be removed in the future and should not be relied on in production.
+        int value = parse_env<int>("TT_METAL_CQ_ASYNC_MODE", static_cast<int>(CommandQueueMode::PASSTHROUGH));
         return static_cast<CommandQueue::CommandQueueMode>(value);
     }
 };

--- a/tt_metal/impl/dispatch/command_queue.hpp
+++ b/tt_metal/impl/dispatch/command_queue.hpp
@@ -29,7 +29,7 @@ using std::tuple;
 using std::unique_ptr;
 
 // Only contains the types of commands which are enqueued onto the device
-enum class EnqueueCommandType { ENQUEUE_READ_BUFFER, ENQUEUE_WRITE_BUFFER, ENQUEUE_PROGRAM, FINISH, ENQUEUE_WRAP, ENQUEUE_RESTART, INVALID };
+enum class EnqueueCommandType { ENQUEUE_READ_BUFFER, ENQUEUE_WRITE_BUFFER, ENQUEUE_PROGRAM, FINISH, ENQUEUE_WRAP, ENQUEUE_RESTART, FLUSH, INVALID };
 
 string EnqueueCommandTypeToString(EnqueueCommandType ctype);
 
@@ -473,7 +473,7 @@ class HWCommandQueue {
 
 
 // Common interface for all command queue types
-struct CommandQueueInterface {
+struct CommandInterface {
     EnqueueCommandType type;
     std::optional<bool> blocking;
     std::optional<std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>>> buffer;
@@ -502,11 +502,14 @@ class CommandQueue {
 
     // Schedule a command to be run on the device
     // Blocking if in passthrough mode. Non-blocking if in async mode
-    void run_command(const CommandQueueInterface& command);
+    void run_command(const CommandInterface& command);
 
     // API for setting the mode of the command queue
     // The environment env counterpart is `TT_METAL_ASYNC_QUEUES`
     void set_mode(const CommandQueueMode& mode);
+
+    // Reference to the underlying hardware command queue, non-const because side-effects are allowed
+    HWCommandQueue& hw_command_queue();
 
    private:
     enum class CommandQueueState {
@@ -514,21 +517,26 @@ class CommandQueue {
         RUNNING = 1,
         TERMINATE = 2,
     };
-    CommandQueueMode mode;
+    CommandQueueMode mode = CommandQueue::get_mode();
     std::unique_ptr<std::thread> worker_thread;
     CommandQueueState worker_state = CommandQueueState::IDLE;
 
-    LockFreeQueue<CommandQueueInterface> worker_queue;
+    LockFreeQueue<CommandInterface> worker_queue;
     uint32_t cq_id;
     Device* device_ptr;
 
     void start_worker();
     void stop_worker();
     void run_worker();
-    void run_command_impl(const CommandQueueInterface& command);
+    void run_command_impl(const CommandInterface& command);
 
     bool async_mode() { return this->mode == CommandQueueMode::ASYNC; }
     bool passthrough_mode() { return this->mode == CommandQueueMode::PASSTHROUGH; }
+
+    static CommandQueueMode get_mode() {
+        int value = parse_env<int>("TT_METAL_ASYNC_QUEUES", static_cast<int>(CommandQueueMode::PASSTHROUGH));
+        return static_cast<CommandQueue::CommandQueueMode>(value);
+    }
 };
 
 } // namespace tt::tt_metal

--- a/tt_metal/impl/dispatch/command_queue.hpp
+++ b/tt_metal/impl/dispatch/command_queue.hpp
@@ -12,13 +12,11 @@
 #include <fstream>
 
 #include "tt_metal/impl/dispatch/command_queue_interface.hpp"
-#include "jit_build/build.hpp"
+#include "tt_metal/impl/dispatch/lock_free_queue.hpp"
 #include "tt_metal/common/base.hpp"
-#include "tt_metal/common/tt_backend_api_types.hpp"
 #include "tt_metal/impl/program/program.hpp"
+#include "common/env_lib.hpp"
 #include "noc/noc_parameters.h"
-#include "third_party/taskflow/taskflow/taskflow.hpp"
-#include "common/executor.hpp"
 #include "tt_metal/host_api.hpp"
 
 namespace tt::tt_metal {
@@ -466,59 +464,60 @@ class HWCommandQueue {
     friend void EnqueueTrace(Trace& trace, bool blocking);
     friend void EndTrace(Trace& trace);
     friend Trace BeginTrace(CommandQueue& cq);
-    friend void EnqueueProgram(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace);
-    friend void EnqueueReadBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking);
-    friend void EnqueueWriteBuffer(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer,
-                                          const void* src, bool blocking);
-    friend void Finish(CommandQueue & cq);
+    friend void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace);
+    friend void EnqueueReadBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking);
+    friend void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, const void* src, bool blocking);
+    friend void FinishImpl(CommandQueue & cq);
     friend class Trace;
 };
 
-class CommandQueue
-{
-    public:
-        CommandQueue () = delete;
-        CommandQueue ( Device * device, uint32_t id) : device_ptr(device), cq_id(id){}
-        ~CommandQueue() {}
 
-        template < class F >
-        tf::AsyncTask& submit ( F && func, std::reference_wrapper< std::shared_future< void > > event ){
-            std::lock_guard<std::mutex> lk(this->mutex);
-            std::tie(this->last, event.get()) = this->last.empty() ? detail::GetExecutor().dependent_async ( std::forward<F>(func) ) : detail::GetExecutor().dependent_async ( std::forward<F>(func), this->last);
-            return this->last;
-        }
+// Common interface for all command queue types
+struct CommandQueueInterface {
+    EnqueueCommandType type;
+    std::optional<bool> token;
+    std::optional<bool> blocking;
+    std::optional<std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>>> buffer;
+    std::optional<std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>>> program;
+    std::optional<const void*> src;
+    std::optional<void*> dst;
+    std::optional<std::reference_wrapper<Trace>> trace;
+};
 
-        //WARNING: Exceptions thrown by sub-tasks are retieved via future; the silent versions don't return a future object, and therefore exceptions are lost
-        template < class F >
-        void submit ( F && func){
-            std::lock_guard<std::mutex> lk(this->mutex);
-            this->last = this->last.empty() ? detail::GetExecutor().silent_dependent_async(std::forward<F>(func) ) : detail::GetExecutor().silent_dependent_async( std::forward<F>(func), this->last );
-        }
+class CommandQueue {
+   public:
+    CommandQueue() = delete;
+    CommandQueue(Device* device, uint32_t id);
+    ~CommandQueue();
 
-        // template < class F >
-        // auto& submit ( F && func){
-        //     std::lock_guard<std::mutex> lk(this->mutex);
-        //     this->last = this->last.has_value() ? detail::GetExecutor().silent_dependent_async((func), this->last.value()) : detail::GetExecutor().silent_dependent_async( func );
-        //     return this->last.value();
-        // }
+    Device* device() const { return this->device_ptr; }
+    uint32_t id() const { return this->cq_id; }
 
-        Device* device() const {
-            return this->device_ptr;
-        }
-        uint32_t id() const{
-            return this->cq_id;
-        }
+    void wait_until_empty();
 
-        void reset() {
-            std::lock_guard<std::mutex> lk(this->mutex);
-            this->last.reset();
-        }
+    void run_command(const CommandQueueInterface& command);
 
-    private:
-        std::mutex mutex;
-        tf::AsyncTask last;
-        uint32_t cq_id;
-        Device * device_ptr;
+   private:
+    enum class CommandQueueState {
+        IDLE = 0,
+        RUNNING = 1,
+        TERMINATE = 2,
+    };
+    bool async_mode = tt::parse_env("TT_METAL_ASYNC_QUEUES", false);
+
+    std::unique_ptr<std::thread> worker_thread;
+    CommandQueueState worker_state = CommandQueueState::IDLE;
+
+    LockFreeQueue<CommandQueueInterface> worker_queue;
+    uint32_t cq_id;
+    Device* device_ptr;
+
+    void start_worker();
+    void stop_worker();
+    void run_worker();
+    void run_command_impl(const CommandQueueInterface& command);
 };
 
 } // namespace tt::tt_metal
+
+std::ostream& operator<<(std::ostream& os, EnqueueCommandType const& type);

--- a/tt_metal/impl/dispatch/lock_free_queue.hpp
+++ b/tt_metal/impl/dispatch/lock_free_queue.hpp
@@ -1,0 +1,53 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <atomic>
+#include <memory>
+
+template<typename T>
+class LockFreeQueue {
+    private:
+        struct Node {
+            std::shared_ptr<T> data;
+            Node* next;
+        };
+
+        std::atomic<Node*> head;
+        std::atomic<Node*> tail;
+
+        Node* pop_head() {
+            Node* oldHead = head.load();
+            if (oldHead == tail.load()) {
+                return nullptr; // Queue is empty
+            }
+            head.store(oldHead->next);
+            return oldHead;
+        }
+
+    public:
+        LockFreeQueue() : head(new Node), tail(head.load()) {}
+
+        void push(const T& value) {
+            std::shared_ptr<T> newData(std::make_shared<T>(value));
+            Node* newNode = new Node;
+            newNode->data = newData;
+            newNode->next = nullptr;
+            Node* oldTail = tail.exchange(newNode);
+            oldTail->next = newNode;
+        }
+
+        std::shared_ptr<T> pop() {
+            Node* oldHead = pop_head();
+            if (!oldHead) {
+                throw std::runtime_error("Queue is empty");
+            }
+            std::shared_ptr<T> result(oldHead->data);
+            delete oldHead;
+            return result;
+        }
+
+        bool empty() const {
+            return head.load() == tail.load();
+        }
+};

--- a/tt_metal/impl/dispatch/lock_free_queue.hpp
+++ b/tt_metal/impl/dispatch/lock_free_queue.hpp
@@ -5,6 +5,8 @@
 #include <atomic>
 #include <memory>
 
+#include "tt_metal/common/assert.hpp"
+
 template<typename T>
 class LockFreeQueue {
     private:
@@ -39,7 +41,7 @@ class LockFreeQueue {
         std::shared_ptr<T> pop() {
             Node* oldHead = pop_head();
             if (!oldHead) {
-                throw std::runtime_error("Queue is empty");
+                TT_THROW("Queue is empty");
             }
             std::shared_ptr<T> result(oldHead->data);
             delete oldHead;

